File: Android/karaokekit-ui/src/main/java/com/netease/yunxin/kit/karaokekit/ui/viewmodel/KaraokeRoomViewModel.java
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

package com.netease.yunxin.kit.karaokekit.ui.viewmodel;

import android.text.TextUtils;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.netease.yunxin.kit.alog.ALog;
import com.netease.yunxin.kit.entertainment.common.utils.Utils;
import com.netease.yunxin.kit.karaokekit.api.NEKaraokeCallback;
import com.netease.yunxin.kit.karaokekit.api.NEKaraokeChorusActionType;
import com.netease.yunxin.kit.karaokekit.api.NEKaraokeEndReason;
import com.netease.yunxin.kit.karaokekit.api.NEKaraokeKit;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeBatchGiftModel;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeChatTextMessage;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeMember;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeOrderSongModel;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeRoomInfo;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeSeatInfo;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeSeatItem;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeSeatRequestItem;
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeSongModel;
import com.netease.yunxin.kit.karaokekit.ui.NEKaraokeUI;
import com.netease.yunxin.kit.karaokekit.ui.R;
import com.netease.yunxin.kit.karaokekit.ui.chatroom.ChatRoomMsgCreator;
import com.netease.yunxin.kit.karaokekit.ui.helper.SeatHelper;
import com.netease.yunxin.kit.karaokekit.ui.listener.MyKaraokeListener;
import com.netease.yunxin.kit.karaokekit.ui.listener.NEKaraokeCallbackWrapper;
import com.netease.yunxin.kit.karaokekit.ui.model.ApplySeatModel;
import com.netease.yunxin.kit.karaokekit.ui.model.OnSeatModel;
import com.netease.yunxin.kit.karaokekit.ui.utils.KaraokeSeatUtils;
import com.netease.yunxin.kit.karaokekit.ui.utils.KaraokeUIUtils;
import com.netease.yunxin.kit.roomkit.api.model.NERoomConnectType;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class KaraokeRoomViewModel extends ViewModel {

  public static final String TAG = "KaraokeRoomViewModel";
  public static final int CURRENT_SEAT_STATE_IDLE = 0;
  public static final int CURRENT_SEAT_STATE_APPLYING = 1;
  public static final int CURRENT_SEAT_STATE_ON_SEAT = 2;

  MutableLiveData<Integer> currentSeatState = new MutableLiveData<>();

  MutableLiveData<NEKaraokeEndReason> errorData = new MutableLiveData<>();
  MutableLiveData<NERoomConnectType> roomConnectState = new MutableLiveData<>();

  MutableLiveData<CharSequence> chatRoomMsgData = new MutableLiveData<>();
  MutableLiveData<Integer> memberCountData = new MutableLiveData<>();

  MutableLiveData<NEKaraokeBatchGiftModel> rewardData = new MutableLiveData<>();

  MutableLiveData<List<OnSeatModel>> onSeatListData = new MutableLiveData<>();

  MutableLiveData<List<ApplySeatModel>> applySeatListData = new MutableLiveData<>();

  MutableLiveData<Boolean> songListData = new MutableLiveData<>();

  NEKaraokeRoomInfo liveInfoNE = null;

  private final MyKaraokeListener listener =
      new MyKaraokeListener() {

        @Override
        public void onReceiveTextMessage(@NonNull NEKaraokeChatTextMessage message) {
          String content = message.getText();
          ALog.i(TAG, "onReceiveTextMessage :${message.fromNick}");
          chatRoomMsgData.postValue(
              ChatRoomMsgCreator.createText(
                  NEKaraokeUI.getInstance().getApplication(),
                  KaraokeUIUtils.isHost(message.getFromUserUuid()),
                  message.getFromNick(),
                  content));
        }

        @Override
        public void onReceiveBatchGift(@NonNull NEKaraokeBatchGiftModel giftModel) {
          ALog.i(TAG, "onReceiveGift");
          rewardData.postValue(giftModel);
        }

        @Override
        public void onMemberAudioMuteChanged(
            @NonNull NEKaraokeMember member, boolean mute, @Nullable NEKaraokeMember operateBy) {
          ALog.i(TAG, "onMemberAudioMuteChanged onSeatList = $onSeatList");
          List<OnSeatModel> onSeatList = SeatHelper.getInstance().getOnSeatItems();
          if (onSeatList != null) {
            for (OnSeatModel seatModel : onSeatList) {
              if (TextUtils.equals(seatModel.getAccount(), member.getAccount())) {
                seatModel.setMute(mute);
              }
            }
            onSeatListData.postValue(onSeatList);
          }
        }

        @Override
        public void onMemberJoinRoom(@NonNull List<NEKaraokeMember> members) {
          for (NEKaraokeMember member : members) {
            ALog.d(TAG, "onMemberJoinRoom :${member.name}");
            if (!KaraokeUIUtils.isLocalAccount(member.getAccount())) {
              chatRoomMsgData.postValue(ChatRoomMsgCreator.createRoomEnter(member.getName()));
            }
          }
          updateRoomMemberCount();
        }

        @Override
        public void onMemberLeaveRoom(@NonNull List<NEKaraokeMember> members) {
          for (NEKaraokeMember member : members) {
            ALog.d(TAG, "onMemberLeaveRoom :$member.name");
            chatRoomMsgData.postValue(ChatRoomMsgCreator.createRoomExit(member.getName()));
          }
          updateRoomMemberCount();
          getSeatRequestList();
        }

        @Override
        public void onMemberJoinChatroom(@NonNull List<NEKaraokeMember> members) {
          for (NEKaraokeMember member : members) {
            ALog.d(TAG, "onMemberJoinChatroom :$member.name");
          }
        }

        @Override
        public void onMemberLeaveChatroom(@NonNull List<NEKaraokeMember> members) {
          for (NEKaraokeMember member : members) {
            ALog.d(TAG, "onMemberJoinChatroom :$member.name");
          }
        }

        @Override
        public void onSeatRequestSubmitted(int seatIndex, @NonNull String account) {
          if (TextUtils.equals(account, KaraokeSeatUtils.getCurrentUuid())) {
            currentSeatState.postValue(CURRENT_SEAT_STATE_APPLYING);
          }
          if (!TextUtils.isEmpty(KaraokeSeatUtils.getMemberNick(account))) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    KaraokeSeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_on_seat_request)));
          }
          getSeatRequestList();
        }

        @Override
        public void onSeatRequestApproved(
            int seatIndex, @NonNull String account, @NonNull String operateBy) {
          if (TextUtils.equals(account, KaraokeSeatUtils.getCurrentUuid())) {
            currentSeatState.postValue(CURRENT_SEAT_STATE_ON_SEAT);
          }
          if (!KaraokeUIUtils.isHost(account)) {
            if (!TextUtils.isEmpty(KaraokeSeatUtils.getMemberNick(account))) {
              chatRoomMsgData.postValue(
                  ChatRoomMsgCreator.createSeatMessage(
                      KaraokeSeatUtils.getMemberNick(account),
                      Utils.getApp().getString(R.string.karaoke_on_seat)));
            }
          }
          getSeatRequestList();
        }

        @Override
        public void onSeatRequestCancelled(int seatIndex, @NonNull String account) {
          if (TextUtils.equals(account, KaraokeSeatUtils.getCurrentUuid())) {
            currentSeatState.postValue(CURRENT_SEAT_STATE_IDLE);
          }
          if (!TextUtils.isEmpty(KaraokeSeatUtils.getMemberNick(account))) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    KaraokeSeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_cancel_on_seat_request)));
          }
          getSeatRequestList();
        }

        @Override
        public void onSeatRequestRejected(
            int seatIndex, @NonNull String account, @NonNull String operateBy) {
          if (TextUtils.equals(account, KaraokeSeatUtils.getCurrentUuid())) {
            currentSeatState.postValue(CURRENT_SEAT_STATE_IDLE);
          }
          getSeatRequestList();
          if (!TextUtils.isEmpty(KaraokeSeatUtils.getMemberNick(account))) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    KaraokeSeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_on_seat_reject)));
          }
        }

        @Override
        public void onSeatLeave(int seatIndex, @NonNull String account) {
          if (TextUtils.equals(account, KaraokeSeatUtils.getCurrentUuid())) {
            currentSeatState.postValue(CURRENT_SEAT_STATE_IDLE);
          }
          if (!TextUtils.isEmpty(KaraokeSeatUtils.getMemberNick(account))) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    KaraokeSeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_down_seat)));
          }
        }

        @Override
        public void onSeatListChanged(@NonNull List<NEKaraokeSeatItem> seatItems) {
          ALog.i(TAG, "onSeatListChanged seatItems = $seatItems");
          onSeatListData.postValue(KaraokeSeatUtils.transNESeatItem2OnSeatModel(seatItems));
        }

        @Override
        public void onSeatKicked(
            int seatIndex, @NonNull String account, @NonNull String operateBy) {
          if (TextUtils.equals(account, KaraokeSeatUtils.getCurrentUuid())) {
            currentSeatState.postValue(CURRENT_SEAT_STATE_IDLE);
          }
          if (!TextUtils.isEmpty(KaraokeSeatUtils.getMemberNick(account))) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    KaraokeSeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_kickout_seat)));
          }
        }

        @Override
        public void onReceiveChorusMessage(
            @NonNull NEKaraokeChorusActionType actionType, @NonNull NEKaraokeSongModel songModel) {
          switch (actionType) {
            case INVITE: // / 邀请消息
              break;
            case AGREE_INVITE: // 同意邀请消息
              break;
            case READY: // / 准备完成可以开始了
              break;
            case START_SONG:
              if (songModel.getOperator() != null
                  && songModel.getOperator().getUserName() != null) {
                chatRoomMsgData.postValue(
                    ChatRoomMsgCreator.createSongMessage(
                        songModel.getOperator().getUserName(),
                        Utils.getApp().getString(R.string.song_start, songModel.getSongName())));
              }
              break;
            case CANCEL_INVITE: // 取消邀请
              break;
            case PAUSE_SONG: // / 歌曲暂停
              if (songModel.getOperator() != null
                  && songModel.getOperator().getUserName() != null) {
                chatRoomMsgData.postValue(
                    ChatRoomMsgCreator.createSongMessage(
                        songModel.getOperator().getUserName(),
                        Utils.getApp().getString(R.string.song_pause, songModel.getSongName())));
              }
              break;
            case RESUME_SONG: // / 恢复播放
              if (songModel.getOperator() != null
                  && songModel.getOperator().getUserName() != null) {
                chatRoomMsgData.postValue(
                    ChatRoomMsgCreator.createSongMessage(
                        songModel.getOperator().getUserName(),
                        Utils.getApp().getString(R.string.song_restart, songModel.getSongName())));
              }
              break;
            case ABANDON:
              break;
            case END_SONG: // / 结束
              break;
          }
        }

        @Override
        public void onRoomEnded(@NonNull NEKaraokeEndReason reason) {
          errorData.postValue(reason);
        }

        @Override
        public void onRoomConnectStateChanged(@NonNull NERoomConnectType state) {
          roomConnectState.postValue(state);
        }

        @Override
        public void onRtcChannelError(int code) {
          if (code == 30015) {
            errorData.postValue(NEKaraokeEndReason.valueOf("END_OF_RTC"));
          }
        }

        @Override
        public void onOrderedSongListChanged() {
          ALog.i(TAG, "onOrderedSongListChanged");
          songListData.postValue(true);
        }

        @Override
        public void onSongOrdered(NEKaraokeOrderSongModel song) {
          ALog.i(TAG, "onSongOrdered song = $song");
          if (song.getOperatorUser() != null && song.getOperatorUser().getUserName() != null) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSongMessage(
                    song.getOperatorUser().getUserName(),
                    Utils.getApp()
                        .getString(
                            R.string.song_ordered,
                            song.getOrderSongResultDto().getOrderSong().getSongName())));
          }
        }

        @Override
        public void onSongDeleted(NEKaraokeOrderSongModel song) {
          ALog.i(TAG, "onSongDeleted song = $song");
          if (song.getOperatorUser() != null && song.getOperatorUser().getUserName() != null) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSongMessage(
                    song.getOperatorUser().getUserName(),
                    Utils.getApp()
                        .getString(
                            R.string.song_deleted,
                            song.getOrderSongResultDto().getOrderSong().getSongName())));
          }
        }

        @Override
        public void onSongTopped(NEKaraokeOrderSongModel song) {
          ALog.i(TAG, "onSongTopped song = $song");
          if (song.getOperatorUser() != null && song.getOperatorUser().getUserName() != null) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSongMessage(
                    song.getOperatorUser().getUserName(),
                    Utils.getApp()
                        .getString(
                            R.string.song_top,
                            song.getOrderSongResultDto().getOrderSong().getSongName())));
          }
        }

        @Override
        public void onNextSong(NEKaraokeOrderSongModel song) {
          ALog.i(TAG, "onSongNext song = $song");
          if (song.getOperatorUser() != null && song.getOperatorUser().getUserName() != null) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSongMessage(
                    song.getOperatorUser().getUserName(),
                    Utils.getApp().getString(R.string.song_switched)));
          }
        }
      };

  public void getSeatRequestList() {
    NEKaraokeKit.getInstance()
        .getSeatRequestList(
            new NEKaraokeCallback<List<NEKaraokeSeatRequestItem>>() {

              @Override
              public void onFailure(int code, @Nullable String msg) {}

              @Override
              public void onSuccess(
                  @Nullable List<NEKaraokeSeatRequestItem> neKaraokeSeatRequestItems) {
                if (neKaraokeSeatRequestItems != null) {
                  List<ApplySeatModel> applySeatList = new ArrayList<>();
                  for (NEKaraokeSeatRequestItem requestItem : neKaraokeSeatRequestItems) {
                    applySeatList.add(
                        new ApplySeatModel(
                            requestItem.getUser(),
                            requestItem.getUserName(),
                            requestItem.getIcon(),
                            ApplySeatModel.SEAT_STATUS_PRE_ON_SEAT));
                  }
                  SeatHelper.getInstance().setApplySeatList(applySeatList);
                  applySeatListData.postValue(applySeatList);
                }
              }
            });
  }

  void refreshLiveInfo(NEKaraokeRoomInfo liveInfoNE) {
    this.liveInfoNE = liveInfoNE;
  }

  public boolean isAnchor(String fromAccount) {
    return liveInfoNE.getAnchor().getAccount().equals(fromAccount);
  }

  public void initDataOnJoinRoom() {
    NEKaraokeKit.getInstance().addKaraokeListener(listener);
    updateRoomMemberCount();
  }

  void updateRoomMemberCount() {
    memberCountData.postValue(NEKaraokeKit.getInstance().getAllMemberList().size());
  }

  public void updateSeat() {
    NEKaraokeKit.getInstance()
        .getSeatInfo(
            new NEKaraokeCallbackWrapper<NEKaraokeSeatInfo>() {

              @Override
              public void onSuccess(@Nullable NEKaraokeSeatInfo seatInfo) {
                if (seatInfo != null) {

                  List<OnSeatModel> onSeatModelList =
                      KaraokeSeatUtils.transNESeatItem2OnSeatModel(seatInfo.getSeatItems());
                  onSeatListData.postValue(onSeatModelList);

                  List<OnSeatModel> onSeatModels = new ArrayList<>();
                  if (onSeatModelList != null) {
                    for (OnSeatModel onSeatModel : onSeatModelList) {
                      if (onSeatModel.getStatus() == ApplySeatModel.SEAT_STATUS_ON_SEAT) {
                        onSeatModels.add(onSeatModel);
                      }
                    }
                  }
                  Collections.sort(onSeatModels);
                  SeatHelper.getInstance().setOnSeatItems(onSeatModels);
                  currentSeatState.postValue(
                      KaraokeSeatUtils.isCurrentOnSeat()
                          ? CURRENT_SEAT_STATE_ON_SEAT
                          : (KaraokeSeatUtils.isCurrentApplyingSeat()
                              ? CURRENT_SEAT_STATE_APPLYING
                              : CURRENT_SEAT_STATE_IDLE));
                }
              }

              @Override
              public void onError(int code, @Nullable String msg) {
                ALog.e(TAG, "getSeatInfo failed code = " + code + " msg = " + msg);
              }
            });
  }

  @Override
  public void onCleared() {
    super.onCleared();
    NEKaraokeKit.getInstance().removeKaraokeListener(listener);
    SeatHelper.getInstance().destroy();
  }

  public MutableLiveData<Integer> getCurrentSeatState() {
    return currentSeatState;
  }

  public MutableLiveData<NEKaraokeEndReason> getErrorData() {
    return errorData;
  }

  public MutableLiveData<NERoomConnectType> getRoomConnectState() {
    return roomConnectState;
  }

  public MutableLiveData<CharSequence> getChatRoomMsgData() {
    return chatRoomMsgData;
  }

  public MutableLiveData<Integer> getMemberCountData() {
    return memberCountData;
  }

  public MutableLiveData<NEKaraokeBatchGiftModel> getRewardData() {
    return rewardData;
  }

  public MutableLiveData<List<OnSeatModel>> getOnSeatListData() {
    return onSeatListData;
  }

  public MutableLiveData<List<ApplySeatModel>> getApplySeatListData() {
    return applySeatListData;
  }

  public MutableLiveData<Boolean> getSongListData() {
    return songListData;
  }

  public NEKaraokeRoomInfo getLiveInfoNE() {
    return liveInfoNE;
  }

  public MyKaraokeListener getListener() {
    return listener;
  }
}

File: Android/karaokekit-ui/src/main/java/com/netease/yunxin/kit/karaokekit/ui/viewmodel/KaraokeRoomViewModel.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.yunxin.kit.karaokekit.ui.viewmodel

import android.text.TextUtils
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.blankj.utilcode.util.Utils
import com.netease.yunxin.kit.alog.ALog
import com.netease.yunxin.kit.karaokekit.api.NEKaraokeCallback
import com.netease.yunxin.kit.karaokekit.api.NEKaraokeChorusActionType
import com.netease.yunxin.kit.karaokekit.api.NEKaraokeEndReason
import com.netease.yunxin.kit.karaokekit.api.NEKaraokeKit
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeChatTextMessage
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeGiftModel
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeMember
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeOrderSongModel
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeRoomInfo
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeSeatItem
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeSeatRequestItem
import com.netease.yunxin.kit.karaokekit.api.model.NEKaraokeSongModel
import com.netease.yunxin.kit.karaokekit.ui.NEKaraokeUI
import com.netease.yunxin.kit.karaokekit.ui.R
import com.netease.yunxin.kit.karaokekit.ui.chatroom.ChatRoomMsgCreator
import com.netease.yunxin.kit.karaokekit.ui.helper.SeatHelper
import com.netease.yunxin.kit.karaokekit.ui.listener.MyKaraokeListener
import com.netease.yunxin.kit.karaokekit.ui.model.ApplySeatModel
import com.netease.yunxin.kit.karaokekit.ui.model.OnSeatModel
import com.netease.yunxin.kit.karaokekit.ui.utils.KaraokeUtils
import com.netease.yunxin.kit.karaokekit.ui.utils.SeatUtils

class KaraokeRoomViewModel : ViewModel() {

    companion object {
        const val TAG = "KaraokeRoomViewModel"
        const val CURRENT_SEAT_STATE_IDLE = 0
        const val CURRENT_SEAT_STATE_APPLYING = 1
        const val CURRENT_SEAT_STATE_ON_SEAT = 2
    }

    var currentSeatState = MutableLiveData<Int>()

    val errorData = MutableLiveData<NEKaraokeEndReason>()

    val chatRoomMsgData = MutableLiveData<CharSequence>()

    val memberCountData = MutableLiveData<Int>()

    val userAccountData = MutableLiveData<Int>()

    val rewardData = MutableLiveData<NEKaraokeGiftModel>()

    val audioEffectFinishData = MutableLiveData<Int>()

    val audioMixingFinishData = MutableLiveData<Boolean>()

    val kickedOutData = MutableLiveData<Boolean>()

    val onSeatListData = MutableLiveData<List<OnSeatModel>>()

    val applySeatListData = MutableLiveData<List<ApplySeatModel>>()

    var liveInfoNE: NEKaraokeRoomInfo? = null

    val songListData = MutableLiveData<Boolean>()

    val switchSongData = MutableLiveData<NEKaraokeOrderSongModel>()

    val nextSongData = MutableLiveData<NEKaraokeOrderSongModel>()

    private var listener = object : MyKaraokeListener() {

        override fun onReceiveTextMessage(message: NEKaraokeChatTextMessage) {
            val content = message.text
            ALog.d(TAG, "onRecvRoomTextMsg :${message.fromNick}")
            chatRoomMsgData.postValue(ChatRoomMsgCreator.createText(NEKaraokeUI.getInstance().application, KaraokeUtils.isHost(message.fromUserUuid), message.fromNick, content))
        }

        override fun onReceiveGift(rewardMsg: NEKaraokeGiftModel) {
            rewardData.postValue(rewardMsg)
        }

        override fun onMemberAudioMuteChanged(
            member: NEKaraokeMember,
            mute: Boolean,
            operateBy: NEKaraokeMember?
        ) {
            val onSeatList = SeatHelper.getInstance().onSeatItems
            ALog.i(TAG, "onMemberAudioMuteChanged onSeatList = $onSeatList")
            onSeatList?.forEach {
                if (TextUtils.equals(it.account, member.account)) {
                    it.isMute = mute
                }
            }
            onSeatListData.postValue(onSeatList)
        }

        override fun onMemberJoinRoom(members: List<NEKaraokeMember>) {
            for (member in members) {
                ALog.d(TAG, "onMemberJoinRoom :${member.name}")
                if (!KaraokeUtils.isMySelf(member.account)) {
                    chatRoomMsgData.postValue(ChatRoomMsgCreator.createRoomEnter(member.name))
                }
            }
            updateRoomMemberCount()
        }

        override fun onMemberLeaveRoom(members: List<NEKaraokeMember>) {
            for (member in members) {
                ALog.d(TAG, "onUserLeft :$member.name")
                chatRoomMsgData.postValue(ChatRoomMsgCreator.createRoomExit(member.name))
            }
            updateRoomMemberCount()
            getSeatRequestList()
        }

        override fun onSeatRequestSubmitted(seatIndex: Int, account: String) {
            if (TextUtils.equals(account, SeatUtils.getCurrentUuid())) {
                currentSeatState.postValue(CURRENT_SEAT_STATE_APPLYING)
            }

            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    SeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_on_seat_request)
                )
            )
            getSeatRequestList()
        }

        override fun onSeatRequestApproved(seatIndex: Int, account: String, operateBy: String) {
            if (TextUtils.equals(account, SeatUtils.getCurrentUuid())) {
                currentSeatState.postValue(CURRENT_SEAT_STATE_ON_SEAT)
            }
            if (!KaraokeUtils.isHost(account)) {
                chatRoomMsgData.postValue(
                    ChatRoomMsgCreator.createSeatMessage(
                        SeatUtils.getMemberNick(account),
                        Utils.getApp().getString(R.string.karaoke_on_seat)
                    )
                )
            }
            getSeatRequestList()
        }

        override fun onSeatRequestCancelled(seatIndex: Int, account: String) {
            if (TextUtils.equals(account, SeatUtils.getCurrentUuid())) {
                currentSeatState.postValue(CURRENT_SEAT_STATE_IDLE)
            }
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    SeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_cancel_on_seat_request)
                )
            )
            getSeatRequestList()
        }

        override fun onSeatRequestRejected(seatIndex: Int, account: String, operateBy: String) {
            if (TextUtils.equals(account, SeatUtils.getCurrentUuid())) {
                currentSeatState.postValue(CURRENT_SEAT_STATE_IDLE)
            }
            getSeatRequestList()
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    SeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_on_seat_reject)
                )
            )
        }

        override fun onSeatLeave(seatIndex: Int, account: String) {
            if (TextUtils.equals(account, SeatUtils.getCurrentUuid())) {
                currentSeatState.postValue(CURRENT_SEAT_STATE_IDLE)
            }
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    SeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_down_seat)
                )
            )
        }

        override fun onSeatListChanged(seatItems: List<NEKaraokeSeatItem>) {
            ALog.i(TAG, "onSeatListChanged seatItems = $seatItems")
            onSeatListData.postValue(SeatUtils.transNESeatItem2OnSeatModel(seatItems))
        }

        override fun onSeatKicked(seatIndex: Int, account: String, operateBy: String) {
            if (TextUtils.equals(account, SeatUtils.getCurrentUuid())) {
                currentSeatState.postValue(CURRENT_SEAT_STATE_IDLE)
            }
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSeatMessage(
                    SeatUtils.getMemberNick(account),
                    Utils.getApp().getString(R.string.karaoke_kickout_seat)
                )
            )
        }

        override fun onReceiveChorusMessage(
            actionType: NEKaraokeChorusActionType,
            songModel: NEKaraokeSongModel
        ) {
            when {
                actionType === NEKaraokeChorusActionType.INVITE -> { // / 邀请消息
                }
                actionType === NEKaraokeChorusActionType.AGREE_INVITE -> { // 同意邀请消息
                }
                actionType === NEKaraokeChorusActionType.READY -> { // / 准备完成可以开始了
                }
                actionType === NEKaraokeChorusActionType.START_SONG -> {
                    chatRoomMsgData.postValue(
                        ChatRoomMsgCreator.createSongMessage(
                            songModel.operator?.userName ?: "",
                            Utils.getApp().getString(R.string.song_start, songModel.songName)
                        )
                    )
                }
                actionType === NEKaraokeChorusActionType.CANCEL_INVITE -> { // 取消邀请
                }
                actionType === NEKaraokeChorusActionType.PAUSE_SONG -> { // / 歌曲暂停
                    chatRoomMsgData.postValue(
                        ChatRoomMsgCreator.createSongMessage(
                            songModel.operator?.userName ?: "",
                            Utils.getApp().getString(R.string.song_pause, songModel.songName)
                        )
                    )
                }
                actionType === NEKaraokeChorusActionType.RESUME_SONG -> { // / 恢复播放
                    chatRoomMsgData.postValue(
                        ChatRoomMsgCreator.createSongMessage(
                            songModel.operator?.userName ?: "",
                            Utils.getApp().getString(R.string.song_restart, songModel.songName)
                        )
                    )
                }
                actionType === NEKaraokeChorusActionType.ABANDON -> {
                }
                actionType === NEKaraokeChorusActionType.END_SONG -> { // / 结束
                }
            }
        }

        override fun onRoomEnded(reason: NEKaraokeEndReason) {
            errorData.postValue(reason)
        }

        override fun onRtcChannelError(code: Int) {
            if (code == 30015) {
                errorData.postValue(NEKaraokeEndReason.fromValue("END_OF_RTC"))
            }
        }

        override fun onSongListChanged() {
            songListData.postValue(true)
        }

        override fun onSongOrdered(song: NEKaraokeOrderSongModel) {
            ALog.i(TAG, "onSongOrder song = $song")
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSongMessage(
                    song.operator?.userName ?: "",
                    Utils.getApp().getString(R.string.song_ordered, song.songName)
                )
            )
        }

        override fun onSongDeleted(song: NEKaraokeOrderSongModel) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSongMessage(
                    song.operator?.userName ?: "",
                    Utils.getApp().getString(R.string.song_deleted, song.songName)
                )
            )
        }

        override fun onSongTopped(song: NEKaraokeOrderSongModel) {
            chatRoomMsgData.postValue(
                ChatRoomMsgCreator.createSongMessage(
                    song.operator?.userName ?: "",
                    Utils.getApp().getString(R.string.song_top, song.songName)
                )
            )
        }

        override fun onNextSong(song: NEKaraokeOrderSongModel, isManual: Boolean) {
            ALog.i(TAG, "onSongNext song = $song")
            if (isManual) {
                chatRoomMsgData.postValue(
                    ChatRoomMsgCreator.createSongMessage(
                        song.operator?.userName ?: "",
                        Utils.getApp().getString(R.string.song_switched, song.songName)
                    )
                )
                switchSongData.postValue(song)
            } else {
                nextSongData.postValue(song)
            }
        }
    }

    fun getSeatRequestList() {
        NEKaraokeKit.getInstance().getSeatRequestList(object : NEKaraokeCallback<List<NEKaraokeSeatRequestItem>> {
            override fun onSuccess(t: List<NEKaraokeSeatRequestItem>?) {
                val applySeatList = mutableListOf<ApplySeatModel>()
                t?.forEach {
                    applySeatList.add(ApplySeatModel(it.user, it.userName, it.icon, ApplySeatModel.SEAT_STATUS_PRE_ON_SEAT))
                }
                SeatHelper.getInstance().applySeatList = applySeatList
                applySeatListData.postValue(applySeatList)
            }

            override fun onFailure(code: Int, msg: String?) {
            }
        })
    }

    fun refreshLiveInfo(liveInfoNE: NEKaraokeRoomInfo) {
        this.liveInfoNE = liveInfoNE
    }

    private fun isAnchor(fromAccount: String): Boolean {
        return liveInfoNE?.anchor?.account == fromAccount
    }

    fun initDataOnJoinRoom() {
        NEKaraokeKit.getInstance().addKaraokeListener(listener)
        updateRoomMemberCount()
    }

    fun updateRoomMemberCount() {
        memberCountData.postUpdateValue {
            NEKaraokeKit.getInstance().allMemberList.size
        }
    }

    override fun onCleared() {
        super.onCleared()
        NEKaraokeKit.getInstance().removeKaraokeListener(listener)
    }
}

internal inline fun <T> MutableLiveData<T>.updateValue(computation: () -> T) {
    value = computation()
}

internal inline fun <T> MutableLiveData<T>.postUpdateValue(computation: () -> T) {
    postValue(computation())
}
